// firebase-tools/lib/deploy/functions/release/lifecycle.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.determineDeploymentDelta = determineDeploymentDelta;
exports.executeLifecycleHooks = executeLifecycleHooks;
const backend = require("../backend");
const error_1 = require("../../../error");
const logger_1 = require("../../../logger");
const utils_1 = require("../../../utils");
const cloudtasks = require("../../../gcp/cloudtasks");
const computeEngine = require("../../../gcp/computeEngine");
const projects_1 = require("../../../management/projects");
const functional_1 = require("../../../functional");
function determineDeploymentDelta(haveBackend) {
    const hasExistingEndpoints = backend.someEndpoint(haveBackend, () => true);
    if (!hasExistingEndpoints) {
        return "afterInstall";
    }
    return "afterUpdate";
}
async function executeLifecycleHooks(wantBackend, haveBackend, plan, codebase) {
    const delta = determineDeploymentDelta(haveBackend);
    const hooks = wantBackend.lifecycleHooks || {};
    const hook = hooks[delta];
    if (!hook) {
        logger_1.logger.debug(`No lifecycle hook configured for event: ${delta}`);
        return false;
    }
    if (delta === "afterUpdate" && plan) {
        const relevantChangesets = Object.entries(plan)
            .filter(([key]) => {
            if (!codebase)
                return true;
            if (key.startsWith(`${codebase}-`))
                return true;
            if (codebase === "default" && key.startsWith("-"))
                return true;
            return false;
        })
            .map(([, c]) => c);
        const hasResourceModifications = relevantChangesets.some((changeset) => changeset.endpointsToCreate.length > 0 ||
            changeset.endpointsToUpdate.length > 0 ||
            changeset.endpointsToDelete.length > 0);
        if (!hasResourceModifications) {
            (0, utils_1.logLabeledBullet)("functions", "No resources modified in codebase. Skipping afterUpdate lifecycle hook.");
            return false;
        }
    }
    if ("task" in hook) {
        (0, utils_1.logLabeledBullet)("functions", `Executing ${delta} lifecycle hook targeting: ${hook.task.function}...`);
        await executeTaskQueueHook(hook.task, wantBackend);
        return true;
    }
    else if ("call" in hook) {
        throw new error_1.FirebaseError(`Lifecycle hook action type "call" is not supported.`);
    }
    else if ("http" in hook) {
        throw new error_1.FirebaseError(`Lifecycle hook action type "http" is not supported.`);
    }
    else {
        (0, functional_1.assertExhaustive)(hook);
    }
}
async function executeTaskQueueHook(taskHook, wantBackend) {
    const targetEndpoint = findTargetEndpoint(wantBackend, taskHook.function);
    if (!targetEndpoint) {
        throw new error_1.FirebaseError(`Target endpoint "${taskHook.function}" not found in backend for lifecycle hook.`);
    }
    if (!backend.isTaskQueueTriggered(targetEndpoint)) {
        throw new error_1.FirebaseError(`Target endpoint "${taskHook.function}" is not a task queue function.`);
    }
    const queueName = cloudtasks.queueNameForEndpoint(targetEndpoint);
    const bodyStr = taskHook.body ? JSON.stringify(taskHook.body) : "";
    const body = bodyStr ? Buffer.from(bodyStr).toString("base64") : undefined;
    const url = targetEndpoint.uri;
    if (!url) {
        throw new error_1.FirebaseError(`Target endpoint "${taskHook.function}" does not have a trigger URI.`);
    }
    const projectMetadata = await (0, projects_1.getProject)(targetEndpoint.project);
    const projectNumber = projectMetadata.projectNumber;
    const sa = targetEndpoint.serviceAccount || (await computeEngine.getDefaultServiceAccount(projectNumber));
    const task = {
        httpRequest: {
            url,
            httpMethod: "POST",
            headers: {
                "Content-Type": "application/json",
            },
            oidcToken: {
                serviceAccountEmail: sa,
                audience: url,
            },
        },
    };
    if (body) {
        task.httpRequest.body = body;
    }
    try {
        await cloudtasks.enqueueTask(queueName, task);
        (0, utils_1.logLabeledSuccess)("functions", `Successfully queued task for lifecycle hook ${taskHook.function} in queue ${queueName}.`);
        (0, utils_1.logLabeledBullet)("functions", `View logs for ${taskHook.function} at: ${getCloudConsoleLogUrl(targetEndpoint)}`);
    }
    catch (err) {
        const errorMsg = err instanceof Error ? err.message : String(err);
        (0, utils_1.logLabeledWarning)("functions", `Failed to enqueue task for lifecycle hook ${taskHook.function}: ${errorMsg}`);
    }
}
function findTargetEndpoint(backendSpec, targetId) {
    for (const endpoint of backend.allEndpoints(backendSpec)) {
        if (endpoint.id === targetId) {
            return endpoint;
        }
    }
    return undefined;
}
function getCloudConsoleLogUrl(endpoint) {
    const { project, region, id } = endpoint;
    const serviceName = endpoint.runServiceId || id;
    const query = `resource.type="cloud_run_revision"\nresource.labels.service_name="${serviceName}"\nresource.labels.location="${region}"`;
    return `https://console.cloud.google.com/logs/query;query=${encodeURIComponent(query)};project=${project}`;
}
